### PR TITLE
Gracefully exit the program when the lease expired

### LINKED ISSUE #2654: KeepAlive in etcd client cannot ensure the permanent validity of lease
`KeepAlive` in etcd client cannot ensure the permanent validity of lease. If automatic renewal of `trillian` lease is interrupted due to unexpected circumstances, trillian will still continue operating. However, the program will never be detected by ETCD server, and as a result it can no longer provide services. In addition, external process health check program cannot discover abnormalities and warnings cannot be sent to system administrators in time.

```golang
// AnnounceSelf announces this binary's presence to etcd.
func AnnounceSelf(ctx context.Context, client *clientv3.Client, etcdService, endpoint string) func() {
	if client == nil {
		return func() {}
	}

	// Get a lease so our entry self-destructs.
	leaseRsp, err := client.Grant(ctx, 30)
	if err != nil {
		glog.Exitf("Failed to get lease from etcd: %v", err)
	}
	client.KeepAlive(ctx, leaseRsp.ID)
...
```

```golang
// recvKeepAlive updates a lease based on its LeaseKeepAliveResponse
func (l *lessor) recvKeepAlive(resp *pb.LeaseKeepAliveResponse) {
	karesp := &LeaseKeepAliveResponse{
		ResponseHeader: resp.GetHeader(),
		ID:             LeaseID(resp.ID),
		TTL:            resp.TTL,
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	ka, ok := l.keepAlives[karesp.ID]
	if !ok {
		return
	}

	if karesp.TTL <= 0 {
		// lease expired; close all keep alive channels
		delete(l.keepAlives, karesp.ID)
		ka.close()
		return
	}
...
```

Meanwhile, because `trillian` did not consume `LeaseKeepAliveResponse` channel immediately, ETCD client would print a large number of warnings #2249 .

### PR DESCRIPTION
**Pull request by px3303**
This PR can let `trillian` proactively Listen "LeaseKeepAliveResponse" channel returned by `KeepAlive` in ETCD client. When automatic renewal interruption is detected,  Exit the program by canceling the context.

Fixes #2654,#2249

### Checklist

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
- [ ] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

### CODE REVIEW

**Comment by mhutchinson** [comment]
Sorry for the slow reply, I've been at work trying to understand this change. The pattern you're using here feels a bit opaque:

1) Code calls a sub-method, which returns a child context
2) Code then awaits cancellation of child context
3) On cancellation of the child context, the parent context is cancelled

I've created a change that I think is equivalent with respect to propagating etcd cancellation, but does it as:

1) Code passes the cancellation function into the sub-method
2) If the lease is cancelled, the cancel function is called

This change is on my fork at https://github.com/google/trillian/compare/master...mhutchinson:etcd_keepalive?expand=1. If you like this, then please do integrate this into the PR. If this change breaks something, then let me know!

**Comment by px3303** [comment]
@mhutchinson 

Sorry for my bad english. Thanks a lot for your suggestion, I decided to go with it. But you removed the `AwaitContext` function, it is needed to exit the rpc server block, So i added it again.

File: cmd/internal/serverutil/main.go [hunk 1]
```diff
@@ -238,3 +248,23 @@ func AnnounceSelf(ctx context.Context, client *clientv3.Client, etcdService, end
 		client.Revoke(ctx, leaseRsp.ID)
 	}
 }
+
+// listenKeepAliveRsp listens to `keepAliveRspCh` channel, and calls the cancel function
+// to notify the lease expired.
+func listenKeepAliveRsp(ctx context.Context, keepAliveRspCh <-chan *clientv3.LeaseKeepAliveResponse, cancel func()) {
```
Comment (AlCutter) [inline cmd/internal/serverutil/main.go:254]: Do you think it'd be slightly easier to grok if this method blocked and the caller is responsible for doing `go listenKeepAliveRsp(...)`?

File: util/process.go [hunk 2]
```diff
@@ -42,3 +42,21 @@ func AwaitSignal(ctx context.Context, doneFn func()) {
 		glog.Infof("AwaitSignal canceled: %v", ctx.Err())
 	}
 }
+
+// AwaitContext waits for context done, then runs the given function.
```
Comment (AlCutter) [inline util/process.go:46]: I'd be tempted to keep this as an unexported function in the `serverutil` package it's called from...

Comment (px3303) [reply util/process.go]: fixed

File: cmd/internal/serverutil/main.go (hunk 1 above)
Comment (px3303) [reply cmd/internal/serverutil/main.go:254]: fixed

**Review by AlCutter — APPROVED** [review]
@mhutchinson for a 2nd pair of eyes

**Review by mhutchinson — APPROVED** [review]

**Pull request merged**